pkg/controller/remotecluster/keystore/keystore.go

```diff
@@ -0,0 +1,293 @@
+		aks.update(pendingChange.remoteClusterName, pendingChange.remoteClusterNamespace, pendingChange.alias, pendingChange.key.keyID, pendingChange.key.encodedValue)
+	}
+
+	if pendingAdds > 0 || pendingDeletions > 0 {
+		aks.log.Info("Pending changes in API keystore", "add", pendingAdds, "deletion", pendingDeletions)
+	}
+	return aks
+}
+
+func (aks *APIKeyStore) Update(remoteClusterName, remoteClusterNamespace, alias, keyID, encodedKeyValue string) *APIKeyStore {
+	// Save the change in memory
+	aks.pendingChanges.AddKey(remoteClusterName, remoteClusterNamespace, alias, keyID, encodedKeyValue)
+	// Load the change in this instance of the store
+	aks.update(remoteClusterName, remoteClusterNamespace, alias, keyID, encodedKeyValue)
+	return aks
+}
+
+func (aks *APIKeyStore) update(remoteClusterName, remoteClusterNamespace, alias, keyID, encodedKeyValue string) {
+	if aks.aliases == nil {
+		aks.aliases = make(map[string]AliasValue)
+	}
+	aks.aliases[alias] = AliasValue{
+		Namespace: remoteClusterNamespace,
+		Name:      remoteClusterName,
+		ID:        keyID,
+	}
+	if aks.encodedKeys == nil {
+		aks.encodedKeys = make(map[string]string)
+	}
+	aks.encodedKeys[alias] = encodedKeyValue
+}
+
+func (aks *APIKeyStore) Aliases() []string {
+	if aks == nil {
+		return nil
+	}
+	aliases := make([]string, len(aks.aliases))
+	i := 0
+	for alias := range aks.aliases {
+		aliases[i] = alias
+		i++
+	}
+	return aliases
+}
+
+func (aks *APIKeyStore) Delete(alias string) *APIKeyStore {
+	// Save the change in memory
+	aks.pendingChanges.DeleteAlias(alias)
+	// Load the change in this instance of the store
+	aks.delete(alias)
+	return aks
+}
+
+func (aks *APIKeyStore) delete(alias string) *APIKeyStore {
+	delete(aks.aliases, alias)
+	delete(aks.encodedKeys, alias)
+	return aks
+}
+
+const (
+	credentialsKeyFormat = "cluster.remote.%s.credentials"
+)
+
+// Save sync the in memory content of the API keystore into the Secret.
+func (aks *APIKeyStore) Save(ctx context.Context, c k8s.Client, owner *esv1.Elasticsearch) error {
+	secretName := types.NamespacedName{
+		Name:      esv1.RemoteAPIKeysSecretName(owner.Name),
+		Namespace: owner.Namespace,
+	}
+	if aks.IsEmpty() {
+		return aks.deleteSecret(ctx, c, secretName)
+	}
+
+	aliases, err := json.Marshal(aks.aliases)
+	if err != nil {
+		return err
+	}
+	data := make(map[string][]byte, len(aks.encodedKeys))
+	for k, v := range aks.encodedKeys {
+		data[fmt.Sprintf(credentialsKeyFormat, k)] = []byte(v)
+	}
+	expectedLabels := labels.AddCredentialsLabel(label.NewLabels(k8s.ExtractNamespacedName(owner)))
+	expectedLabels[commonv1.TypeLabelName] = RemoteClusterAPIKeysType
+	expected := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName.Name,
+			Namespace: secretName.Namespace,
+			Annotations: map[string]string{
+				aliasesAnnotationName: string(aliases),
+			},
+			Labels: expectedLabels,
+		},
+		Data: data,
+	}
+	if _, err := reconciler.ReconcileSecret(ctx, c, expected, owner); err != nil {
```

Comment:
I was seeing a few noisy conflicts during testing I assume because `Save` is called in two places during one reconciliation?

```
2024-10-20T10:29:43.425+0200	ERROR	manager.eck-operator	Reconciler error	{"service.version": "2.15.0-SNAPSHOT+65769109", "controller": "remotecluster-controller", "object": {"name":"cluster1","namespace":"ns1"}, "namespace": "ns1", "name": "cluster1", "reconcileID": "316f64b0-454e-4285-8e89-34c18f0da6a2", "error": "Operation cannot be fulfilled on secrets \"cluster1-es-remote-api-keys\": the object has been modified; please apply your changes to the latest version and try again", "errorCauses": [{"error": "Operation cannot be fulfilled on secrets \"cluster1-es-remote-api-keys\": the object has been modified; please apply your changes to the latest version and try again"}]}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler
	/Users/pebrc/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:316
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem
	/Users/pebrc/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:263
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2
	/Users/pebrc/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:224
```